test/udev_mapping_test.rb

```diff
@@ -42,56 +42,142 @@
+        before do
+          allow(device).to receive(:udev_ids).and_return(ids)
+        end
 
-      expect(subject.to_mountby_device(device.name)).to eq "/dev/disk/by-id/abc"
+        context "and the device ids are known" do
+          let(:ids) { ["abc", "cde"] }
+
+          it "returns the first by id udev link" do
+            expect(subject.to_mountby_device(device.name)).to eq("/dev/disk/by-id/abc")
+          end
+        end
+
+        context "and the device ids are unknown" do
+          let(:ids) { [] }
+
+          it "returns the kernel name as fallback" do
+            expect(subject.to_mountby_device(device.name)).to eq("/dev/sda3")
+          end
+        end
+      end
+
+      context "and the mount by option is by PATH" do
+        let(:mount_by_option) { :path }
+
+        before do
+          allow(device).to receive(:udev_paths).and_return(paths)
+        end
+
+        context "and the device paths are known" do
+          let(:paths) { ["abc", "cde"] }
+
+          it "returns the first by path udev link" do
+            expect(subject.to_mountby_device(device.name)).to eq("/dev/disk/by-path/abc")
+          end
+        end
+
+        context "and the device paths are unknown" do
+          let(:paths) { [] }
+
+          it "returns the kernel name as fallback" do
+            expect(subject.to_mountby_device(device.name)).to eq("/dev/sda3")
+          end
+        end
+      end
     end
 
-    it "returns first udev link by path if defined" do
-      device.filesystem.remove_mount_point
-      allow(device).to receive(:udev_full_label).and_return(nil)
-      allow(device).to receive(:udev_full_uuid).and_return(nil)
-      allow(device).to receive(:udev_paths).and_return(["abc", "cde"])
+    context "when the device is mounted" do
+      before do
+        device.filesystem.mount_point.mount_by = mount_by
+      end
+
+      let(:mount_by_option) { :label }
+
+      it "returns the udev link according to its mount point configuration" do
+        expect(subject.to_mountby_device(device.name)).to eq("/dev/disk/by-label/DATA")
+      end
 
-      expect(subject.to_mountby_device(device.name)).to eq "/dev/disk/by-path/abc"
+      include_examples "options_and_fallback"
     end
 
-    it "returns kernel name as last fallback" do
-      device.filesystem.remove_mount_point
-      allow(device).to receive(:udev_full_label).and_return(nil)
-      allow(device).to receive(:udev_full_uuid).and_return(nil)
+    context "when the device is not mounted" do
+      before do
+        device.filesystem&.remove_mount_point
+
+        allow_any_instance_of(Y2Storage::MountPoint).to receive(:preferred_mount_by)
+          .and_return(mount_by)
+      end
+
+      let(:mount_by_option) { :label }
+
+      it "returns the udev link according to the preferred mount by" do
+        expect(subject.to_mountby_device(device.name)).to eq("/dev/disk/by-label/DATA")
+      end
```

Comment:
I think we should not double test in storage. Simply delegates to storage-ng is enough for me here. As it is out of module bonds.

Reply:
Ok, I will simplify it.